flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,434 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.SimpleCounter;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.streaming.api.operators.InputSelectable;
+import org.apache.flink.streaming.api.operators.InputSelection;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.metrics.WatermarkGauge;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatus;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatusMaintainer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Input reader for {@link org.apache.flink.streaming.runtime.tasks.TwoInputSelectableStreamTask}
+ * in the case that the operator is InputSelectable.
+ *
+ * @param <IN1> The type of the records that arrive on the first input
+ * @param <IN2> The type of the records that arrive on the second input
+ */
+@Internal
+public class StreamTwoInputSelectableProcessor<IN1, IN2> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(StreamTwoInputSelectableProcessor.class);
+
+	private volatile boolean continuousProcessing = true;
+
+	private final NetworkInput input1;
+	private final NetworkInput input2;
+
+	private final Object lock;
+
+	private final TwoInputStreamOperator<IN1, IN2, ?> streamOperator;
+
+	private final InputSelectable inputSelector;
+
+	private final AuxiliaryHandler auxiliaryHandler;
+
+	private final CompletableFuture<Integer>[] listenFutures;
+
+	private final boolean[] isFinished;
+
+	private InputSelection inputSelection;
+
+	private AtomicInteger availableInputsMask = new AtomicInteger();
```

Comment:
**(hot looping without synchronisation part 2)** I would drop `AtomicInteger availableInputsMask` and replace it with `integer availableIntpusMask` which is updated based on `BufferOrEvent#moreAvailable()` mentioned above.

As long as all of the selected inputs are available, there is no need to synchronise. If some selected input is not available, that's the point where we could use `AtomicInteger`, but for `TwoInputStream` (and not general N input) case more or less equally efficient would be to check `listenFutures[x].getNow(-1)`. This is also a single volatile access under the hood, with the same performance penalty as `availableInputsMask.get()`.

But again, keep in mind that while hot looping you can not touch `listenFutures`.

Reply:
When reading with `ANY` mode, there are the following cases.
1. If two inputs are all available, take turns to read them (for fair reading).
2. If only one of the inputs is available, always read it until another input become available or itself becomes unavailable/finished.
3. If neither input is available, block to wait for any input to become available.

Only using `BufferOrEvent#moreAvailable()` to update `int availableIntpusMask`, I think that there are some problems with the second case. Assuming that one of the inputs is unavailable at a certain moment and the corresponding bit mask in `availableInputsMask` is set to `0`, there are several possible ways to select which input to read.
1. We read the currently available input until it becomes unavailable or finished, but if the another input becomes available during this period, it will not be able to adjust to fair reading in time.
2. Always read with absolute fairness, that is, reading a record from the available input, then checking and waiting for the unavailable input to become available and reading it, but it is undoubtedly inefficient.

I only  thought of  these ways. I didn't understand that way you said.

Reply:
True, 2. is the worst case and to provide fairness we have to either constantly check if the second input became suddenly available (`listenFutures[x]`) or at least periodically do that, for example once per buffer or something like that. I would start with always and later we can try to optimise it. 

(Old `TwoInputStreamProcessor` was checking/synchronising only once per buffer)

Reply:
> I would start with always and later we can try to optimise it.

That is to say, we are currently taking the first way, right? Does failure to adjust to fair selection in time make the execution of the whole topology (the job) still less efficient than "streamtwoinputprocessor" in case of `ANY` mode ?

Reply:
Yes, in case of 2.:
> If only one of the inputs is available, always read it until another input become available or itself becomes unavailable/finished.

let's start with always checking once per record if the other side of the input became available.

After the holiday (we have now a public holiday in Germany) I'll have to think about this a little bit more, what would be the most general and efficient way how to handle this without this overhead.

Reply:
I got it.  Wish you a good holiday.